### PR TITLE
[backport] Upgrade 2.12.x build to SBT 1.3.7 [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
Work continues on 2.12.x for users yet to upgrade to 2.13.x. Keeping this
branch on increasingly archaic versions of SBT and MiMa carries some risk
and adds friction when switching context between branches.

This commit backports the upgade to SBT 1.2.x (#6256) and the subsequent
upgrade to 1.3.7 (#8525 #8652). This includes updating MiMa to 0.6.1 and moving
the MiMa filter lists into build.sbt and adding exclusions for
additional incompatibilties that the new version of MiMa reports
but that its too late for us to heed.

### CODE REVIEW

**Comment by retronym** [comment]
```
[error] [launcher] error during sbt launcher: sbt/sbt#4955!

[error] [launcher] sbt launcher is unable to detect the Scala version for org.scala-sbt:sbt;

[error] [launcher] this likely indicates an incomplete artifact resolution and/or corrupt boot cache (~/.sbt/boot/).

[error] [launcher] the following is the full classpath that was retrieved for org.scala-sbt:sbt:

[error] [launcher]  * /home/travis/.sbt/boot/other/org.scala-sbt/sbt/1.3.7/sbt-1.3.7.jar
```

sbt/sbt#4955

**Comment by retronym** [comment]
Nuking the travis caches, I now get:

```
1.1.3.pom (Server returned HTTP response code: 403 for URL: https://repo1.maven.org/maven2/org/scala-sbt/launcher-interface/1.1.3/launcher-interface-1.1.3.pom) while downloading https://repo1.maven.org/maven2/org/scala-sbt/launcher-interface/1.1.3/launcher-interface-1.1.3.pom

[error]   not found: /home/travis/.ivy2/local/org.scala-sbt/launcher-interface/1.1.3/ivys/ivy.xml

[error]   not found: https://repo.scala-sbt.org/scalasbt/sbt-plugin-releases/org.scala-sbt/launcher-interface/1.1.3/ivys/ivy.xml

[error]   not found: https://repo.typesafe.com/typesafe/ivy-releases/org.scala-sbt/launcher-interface/1.1.3/ivys/ivy.xml

[error] Error downloading jline:jline:2.14.6

[error]   Not found

[error]   Not found

[error]   download error: Caught java.io.IOException: Server returned HTTP response code: 403 for URL: https://repo1.maven.org/maven2/jline/jline/2.14.6/jline-2.14.6.pom (Server returned HTTP response code: 403 for URL: https://repo1.maven.org/maven2/jline/jline/2.14.6/jline-2.14.6.pom) while downloading https://repo1.maven.org/maven2/jline/jline/2.14.6/jline-2.14.6.pom

[error]   not found: /home/travis/.ivy2/local/jline/jline/2.14.6/ivys/ivy.xml
```

**Comment by dwijnand** [comment]
In backporting was there any particular part you think could use more attention in reviewing?  Otherwise, I'm happy with this landing if it goes green.

"error during sbt launcher" means (not generally, but what it means these days that it's popped up everywhere) I, the launcher, got 403'd by Maven Central downloading the app (sbt).  That's just a risk when upgrading, but if you're not upgrading it means you're not caching the launcher jars (`~/.sbt/launchers`).  As we're upgrading here so we just need to retry and retry the PR (ironically, given the 403 is for rate-limiting...), deleting the cache wasn't necessary (IMO).

**Comment by SethTisue** [comment]
since the 2.13.x PR was merged, our Windows job on Jenkins stopped working: https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-windows/1251/console

let's see if this PR has the same problem: https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-windows/1627/ (queued)

**Comment by dwijnand** [comment]
Perhaps the changes that the community build layers on the build need adapting.  Or maybe the stuff is in dbuild, in some "how to consume fresh scala" part of it (just a guess)?

**Comment by dwijnand** [comment]
> let's see if this PR has the same problem: [scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-windows/1627](https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-windows/1627/) (queued)

failed on smtp.  retry https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-windows/1629/

**Comment by SethTisue** [comment]
the Windows jobs don't use dbuild. see `scripts/jobs/integrate/windows`. the script builds Scala itself, it doesn't consume a Scala built elsewhere (except STARR, I mean)

**Comment by retronym** [comment]
Here's the current blocker: https://github.com/sbt/zinc/pull/571#issuecomment-578467052

A warning for Scaladoc for a use case appears to have an invalid end position, and Zinc has started eagerly converting positions to lines as a convenience to BSP clients. This results in an `ArrayIndexOutOfBoundsException`. We don't see this on the 2.13.x branch as it seems the buggy positions are for `@usecase`-s which aren't needed in the 2.13.x collections docs.

**Comment by retronym** [comment]
> In backporting was there any particular part you think could use more attention in reviewing? Otherwise, I'm happy with this landing if it goes green.

The main risk of this sort of backport is failing to also backport followup changes on that branch. I've been reviewing the full history of `-- build.sbt project/` on 2.13.x to look for such commits. 

I'll also do a byte-for-byte comparison check on the generated artifacts (JARs, ZIPs, docs etc) once the build is green to look for any unexpected changes.

File: src/library/scala/collection/GenIterableLike.scala [hunk 1]
```diff
@@ -147,3 +147,13 @@ trait GenIterableLike[+A, +Repr] extends Any with GenTraversableLike[A, Repr] {
   def zipAll[B, A1 >: A, That](that: GenIterable[B], thisElem: A1, thatElem: B)(implicit bf: CBF[Repr, (A1, B), That]): That
 
 }
+
+// Workaround for Scaladoc @usecase position bug scala/issues#11865
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (SethTisue) [inline src/library/scala/collection/GenIterableLike.scala:152]: 🙈

Comment (retronym) [reply src/library/scala/collection/GenIterableLike.scala:152]: After the [fix](https://github.com/scala/scala/pull/8668) is in an we re-STARR this can go.  👀 🐵

Comment (retronym) [reply src/library/scala/collection/GenIterableLike.scala:152]: Fix in https://github.com/scala/scala/pull/8668, but we'll need the workaround until we re-STARR atop that.

**Comment by retronym** [comment]
I'm happy that we're not seeing diffs in the generated classes, JARs (other than timestamps). Scaladoc looks functionally the same although it appears that the diagram HTML output is not stable from run-to-run which makes the diffing less simple.

**Comment by retronym** [comment]
I've also done my pass through the commits on the build and added three more backports:


 - Restore the working directory of partest, junit to the project root
 - Resort to old-school git HEAD detection in the build
 - Update Scala' build dependencies junit,jol,jackson

File: test/osgi/src/ScalaOsgiHelper.scala [hunk 2]
```diff
@@ -7,6 +7,7 @@ import java.io.File
 trait ScalaOsgiHelper {
 
   private def allBundleFiles = {
+    println(new File(".").getAbsolutePath)
```
Comment (lrytz) [inline test/osgi/src/ScalaOsgiHelper.scala]: remove?

**Comment by retronym** [comment]
I'd vote to defer merging this until after 2.12.11 which is scheduled in a few weeks. What do others think?

**Comment by dwijnand** [comment]
As there's nothing urgent about this, sure.

**Comment by lrytz** [comment]
I'd be fine merging this now. Jardiff was done - on which artifacts? I think the `publishLocal` artifacts include the OSGi stuff, so that would be covered. Maybe we can add a test commit that breaks bin compat to test MiMa.

File: src/library/scala/collection/GenIterableLike.scala (hunk 1 above)
Comment (lrytz) [reply src/library/scala/collection/GenIterableLike.scala:152]: The fix was merged, the de-slash-forestation can take place

**Comment by SethTisue** [comment]
It's true there's nothing urgent about this upgrade, but there's also nothing urgent or unusually high stakes about the 2.12.11 release. I think we might as well go ahead and risk the regression now; if we are doing this at all on 2.12.x, we have to risk it sometime.

TL;DR yup let's merge (though, no strong feeling either way)

**Comment by dwijnand** [comment]
My thinking is we wouldn't want to have any sbt 1.3 issue affecting the 2.12 and perhaps also the 2.13 release, nor be fixed in a rush.  So let's merge right after 2.12.11 and let the development cycle after that be available for fixups.

But we can also just revert this PR if it becomes problematic.

**Comment by retronym** [comment]
Okay, let's pull the trigger.

**Comment by retronym** [comment]
My jardiffing was based on the contents of `build/pack` after `sbt setupPubishCore dist/mkPack`
